File: working hours register/Level 3/C#/worker_service.cs
// WorkerService.cs
using System.Collections.Generic;
using System.Linq;

public class WorkerService
{
    private readonly WorkerRepository _repo;

    public WorkerService(WorkerRepository repo)
    {
        _repo = repo;
    }

    public string AddWorker(string workerId, string position, int compensation)
    {
        return _repo.AddWorker(workerId, position, compensation);
    }

    public string Register(string workerId, int timestamp)
    {
        var worker = _repo.GetWorker(workerId);
        return worker != null ? worker.Register(timestamp) : "invalid_request";
    }

    public string GetTime(string workerId)
    {
        var worker = _repo.GetWorker(workerId);
        return worker != null ? worker.GetTotalMinutes() : "";
    }

    public string Promote(string workerId, string newPosition, int compensation, int startTimestamp)
    {
        var worker = _repo.GetWorker(workerId);
        return worker != null ? worker.Promote(newPosition, compensation, startTimestamp) : "invalid_request";
    }

    public string CalcSalary(string workerId, int start, int end)
    {
        var worker = _repo.GetWorker(workerId);
        return worker != null ? worker.CalcSalary(start, end) : "";
    }

    public string TopNWorkers(int n, string position)
    {
        var workers = _repo.GetWorkersByPosition(position);
        if (workers.Count == 0)
            return "";

        var topN = workers
            .OrderByDescending(w => w.TotalMinutes)
            .ThenBy(w => w.WorkerId)
            .Take(n)
            .Select(w => $"{w.WorkerId}({w.TotalMinutes})");

        return string.Join(", ", topN);
    }
}

File: working hours register/Level 3/Python/worker_service.py
# worker_service.py

class WorkerService:
    def __init__(self, repo: WorkerRepository):
        self.repo = repo

    def add_worker(self, worker_id: str, position: str, compensation: int) -> str:
        return self.repo.add_worker(worker_id, position, compensation)

    def register(self, worker_id: str, timestamp: int) -> str:
        worker = self.repo.get_worker(worker_id)
        return worker.register(timestamp) if worker else "invalid_request"

    def get_time(self, worker_id: str) -> str:
        worker = self.repo.get_worker(worker_id)
        return worker.get_total_minutes() if worker else ""

    def top_n_workers(self, n: int, position: str) -> str:
        workers = self.repo.get_workers_by_position(position)
        if not workers:
            return ""

        top_n = sorted(workers, key=lambda w: (-w.total_minutes, w.worker_id))[:n]
        return ", ".join(f"{w.worker_id}({w.total_minutes})" for w in top_n)

    def promote(self, worker_id: str, position: str, compensation: int, start_ts: int) -> str:
        worker = self.repo.get_worker(worker_id)
        return worker.promote(position, compensation, start_ts) if worker else "invalid_request"

    def calc_salary(self, worker_id: str, start: int, end: int) -> str:
        worker = self.repo.get_worker(worker_id)
        return worker.calc_salary(start, end) if worker else ""
